src/skycoin/skycoin.go

```diff
@@ -368,8 +297,178 @@ func (c *Coin) initLogFile() (*os.File, error) {
+	}
+
+	if dbVersion == nil {
+		c.logger.Info("DB version not found in DB")
+	} else {
+		c.logger.Infof("DB version: %s", dbVersion)
+	}
+
+	c.logger.Infof("DB verify checkpoint version: %s", DBVerifyCheckpointVersion)
+
+	// If the saved DB version is higher than the app version, abort.
+	// Otherwise DB corruption could occur.
+	if dbVersion != nil && dbVersion.GT(*appVersion) {
+		err = fmt.Errorf("Cannot use newer DB version=%v with older software version=%v", dbVersion, appVersion)
+		c.logger.WithError(err).Error()
+		return nil, err
+	}
+
+	// Verify the DB if the version detection says to, or if it was requested on the command line
+	if shouldVerifyDB(appVersion, dbVersion) || c.config.Node.VerifyDB {
+		if c.config.Node.ResetCorruptDB {
+			// Check the database integrity and recreate it if necessary
+			c.logger.Info("Checking database and resetting if corrupted")
+			newDB, err := visor.ResetCorruptDB(db, c.config.Node.blockchainPubkey, quit)
+			if err != nil {
+				if err != visor.ErrVerifyStopped {
+					c.logger.Errorf("visor.ResetCorruptDB failed: %v", err)
+				}
+				return nil, err
+			}
+			db = newDB
+		} else {
+			c.logger.Info("Checking database")
+			if err := visor.CheckDatabase(db, c.config.Node.blockchainPubkey, quit); err != nil {
+				if err != visor.ErrVerifyStopped {
+					c.logger.Errorf("visor.CheckDatabase failed: %v", err)
+				}
+				return nil, err
+			}
+		}
+	}
+
+	// Update the DB version
+	if !db.IsReadOnly() {
+		if err := visor.SetDBVersion(db, *appVersion); err != nil {
+			c.logger.WithError(err).Error("visor.SetDBVersion failed")
+			return nil, err
+		}
+	}
+
+	return db, nil
+}
+
+func shouldVerifyDB(appVersion, dbVersion *semver.Version) bool {
+	// If the dbVersion is not set, verify
+	if dbVersion == nil {
+		return true
+	}
+
+	// If the dbVersion is less than the verification checkpoint version
+	// and the appVersion is greater than or equal to the checkpoint version,
+	// verify
+	if dbVersion.LT(dbVerifyCheckpointVersionParsed) && appVersion.GTE(dbVerifyCheckpointVersionParsed) {
+		return true
+	}
+
+	return false
+}
+
+// initWallet creates new instance of *wallets.Servcie
+func (c *Coin) initWallet() (*wallet.Service, error) {
+	cfg := c.configureWallet()
+
+	s, err := wallet.NewService(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return s, nil
+}
+
+// configureWallet sets the wallet config values
+func (c *Coin) configureWallet() wallet.Config {
+	wc := wallet.NewConfig()
+
+	wc.WalletDir = c.config.Node.WalletDirectory
+	_, wc.EnableWalletAPI = c.config.Node.enabledAPISets[api.EndpointsWallet]
+	_, wc.EnableSeedAPI = c.config.Node.enabledAPISets[api.EndpointsInsecureWalletSeed]
+
+	// Initialize wallet default crypto type
+	cryptoType, err := wallet.CryptoTypeFromString(c.config.Node.WalletCryptoType)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	wc.CryptoType = cryptoType
+
+	return wc
+}
+
+// initVisor creates new instance of *visor.Visor
+func (c *Coin) initVisor(
```

Comment:
Can this be a method in `visor` package?

Reply:
It can be only if it will return underlying instances of UnconfirmedTransactionPooler, Blockhchainer and Historyer.

Look at [this issue](https://github.com/watercompany/obelisk/issues/1l) and attached commit. I need access to instances of those interfaces to override some methods and define new one.

And in that case I can not create, for example, custom Blockchainer with overridden methods and put it into Visor. However, that is not needed as for now.

Reply:
The function can be broken down:

- creating Blockchain, UnconfirmedTxnPool, HistoryDB
- Initializing database contents (unspent pool, history db indexes)

So i think it can be rearranged so that the db.Update remains in the visor package.